filestore/filepath.go

```diff
@@ -0,0 +1,520 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package filestore
+
+import (
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"strings"
+)
+
+type FileType string
+
+type FileStoreType string
+
+const (
+	Memory     FileStoreType = "MEMORY"
+	FileSystem FileStoreType = "LOCAL_FILESYSTEM"
+	Azure      FileStoreType = "AZURE"
+	S3         FileStoreType = "S3"
+	GCS        FileStoreType = "GCS"
+	HDFS       FileStoreType = "HDFS"
+)
+
+const (
+	Parquet FileType = "parquet"
+	CSV     FileType = "csv"
+	DB      FileType = "db"
+)
+
+const (
+	GSPrefix         = "gs://"
+	S3Prefix         = "s3://"
+	S3APrefix        = "s3a://"
+	AzureBlobPrefix  = "abfss://"
+	HDFSPrefix       = "hdfs://"
+	FileSystemPrefix = "file://"
+)
+
+var ValidSchemes = []string{
+	GSPrefix, S3Prefix, S3APrefix, AzureBlobPrefix, HDFSPrefix, FileSystemPrefix,
+}
+
+func (ft FileType) Matches(file string) bool {
+	ext := GetFileExtension(file)
+	return FileType(ext) == ft
+}
+
+func IsValidFileType(file string) bool {
+	for _, fileType := range []FileType{Parquet, CSV, DB} {
+		if fileType.Matches(file) {
+			return true
+		}
+	}
+	return false
+}
+
+func GetFileExtension(file string) string {
+	ext := filepath.Ext(file)
+	return strings.ReplaceAll(ext, ".", "")
+}
+
+type Filepath interface {
+	// Scheme encompasses
+	// * protocol (e.g. s3://, gs://, abfss://)
+	// * host (e.g. <container>@<account>.dfs.core.windows.net for Azure Blob)
+	// * port (if applicable)
+	// This naming technically conflicts with the standard definition of "scheme," which
+	// is only the protocol _without_ the domain and port (i.e. authority); however, given
+	// we're not using these components independently and there's no single term to denote
+	// <SCHEME>://<HOST>:<PORT>, we're using "scheme" to encompass all three.
+	Scheme() string
+	SetScheme(scheme string) error
+
+	// Returns the name of the bucket (S3) or container (Azure Blob Storage)
+	Bucket() string
+	SetBucket(bucket string) error
+
+	// Returns the blob key, which is the relative path to the object (i.e. without the scheme or bucket/container)
+	Key() string
+	SetKey(key string) error
+
+	// Returns the key prefix (i.e. the directory path to the object)
+	KeyPrefix() string
+
+	IsDir() bool
+	SetIsDir(isDir bool)
+
+	// Returns the file extension (e.g. "parquet", "csv", etc. of the object)
+	Ext() FileType
+
+	// Returns the full path to the object, including the scheme and bucket/container
+	ToURI() string
+	// Consumes a URI (e.g. abfss://<container>@<storage_account>/path/to/file) and parses it into
+	// the specific parts that the implementation expects.
+	ParseFilePath(path string) error
+	ParseDirPath(path string) error
+
+	Validate() error
+	IsValid() bool
```

Comment:
Why are we storing whether it's valid or not? I imagine we shouldn't even successfully create a file path if it's not valid?

Reply:
like why should we store a filepath that's not valid?